File: service/src/test/java/org/coner/core/util/HibernateEntityTestUtils.java
package org.coner.core.util;

import java.math.BigDecimal;
import java.time.Year;
import java.util.Date;
import java.util.Set;

import org.coner.core.hibernate.entity.CarHibernateEntity;
import org.coner.core.hibernate.entity.CompetitionGroupHibernateEntity;
import org.coner.core.hibernate.entity.CompetitionGroupSetHibernateEntity;
import org.coner.core.hibernate.entity.EventHibernateEntity;
import org.coner.core.hibernate.entity.HandicapGroupHibernateEntity;
import org.coner.core.hibernate.entity.HandicapGroupSetHibernateEntity;
import org.coner.core.hibernate.entity.PersonHibernateEntity;
import org.coner.core.hibernate.entity.RegistrationHibernateEntity;
import org.coner.core.hibernate.entity.RunHibernateEntity;

import com.google.common.collect.Sets;

public final class HibernateEntityTestUtils {

    private HibernateEntityTestUtils() {
    }

    public static PersonHibernateEntity fullPerson() {
        return fullPerson(
                TestConstants.PERSON_ID,
                TestConstants.PERSON_FIRST_NAME,
                TestConstants.PERSON_MIDDLE_NAME,
                TestConstants.PERSON_LAST_NAME
        );
    }

    public static PersonHibernateEntity fullPerson(String id, String firstName, String middleName, String lastName) {
        PersonHibernateEntity person = new PersonHibernateEntity();
        person.setId(id);
        person.setFirstName(firstName);
        person.setMiddleName(middleName);
        person.setLastName(lastName);
        return person;
    }

    public static CarHibernateEntity fullCar() {
        return fullCar(
                TestConstants.CAR_ID,
                TestConstants.CAR_YEAR,
                TestConstants.CAR_MAKE,
                TestConstants.CAR_MODEL,
                TestConstants.CAR_TRIM,
                TestConstants.CAR_COLOR
        );
    }

    public static CarHibernateEntity fullCar(
            String id,
            Year year,
            String make,
            String model,
            String trim,
            String color
    ) {
        CarHibernateEntity car = new CarHibernateEntity();
        car.setId(id);
        car.setYear(year);
        car.setMake(make);
        car.setModel(model);
        car.setTrim(trim);
        car.setColor(color);
        return car;
    }

    public static EventHibernateEntity fullEvent() {
        return fullEvent(
                TestConstants.EVENT_ID,
                TestConstants.EVENT_NAME,
                TestConstants.EVENT_DATE,
                fullHandicapGroupSet(),
                fullCompetitionGroupSet(),
                TestConstants.EVENT_MAX_RUNS_PER_REGISTRATION,
                TestConstants.EVENT_CURRENT
        );
    }

    public static EventHibernateEntity fullEvent(
            String id,
            String name,
            Date date,
            HandicapGroupSetHibernateEntity handicapGroupSet,
            CompetitionGroupSetHibernateEntity competitionGroupSet,
            int maxRunsPerRegistration,
            boolean current
    ) {
        EventHibernateEntity event = new EventHibernateEntity();
        event.setId(id);
        event.setName(name);
        event.setDate(date);
        event.setHandicapGroupSet(handicapGroupSet);
        event.setCompetitionGroupSet(competitionGroupSet);
        event.setMaxRunsPerRegistration(maxRunsPerRegistration);
        event.setCurrent(current);
        return event;
    }

    public static RegistrationHibernateEntity fullRegistration() {
        return fullRegistration(
                TestConstants.REGISTRATION_ID,
                fullPerson(),
                fullCar(),
                fullEvent(),
                fullHandicapGroup(),
                fullCompetitionGroup(),
                TestConstants.REGISTRATION_NUMBER,
                TestConstants.REGISTRATION_CHECKED_IN
        );
    }

    public static RegistrationHibernateEntity fullRegistration(
            String id,
            PersonHibernateEntity person,
            CarHibernateEntity car,
            EventHibernateEntity event,
            HandicapGroupHibernateEntity handicapGroup,
            CompetitionGroupHibernateEntity competitionGroup,
            String number,
            boolean checkedIn
    ) {
        RegistrationHibernateEntity registration = new RegistrationHibernateEntity();
        registration.setId(id);
        registration.setPerson(person);
        registration.setCar(car);
        registration.setEvent(event);
        registration.setHandicapGroup(handicapGroup);
        registration.setCompetitionGroup(competitionGroup);
        registration.setNumber(number);
        registration.setCheckedIn(checkedIn);
        return registration;
    }

    public static HandicapGroupHibernateEntity fullHandicapGroup() {
        return fullHandicapGroup(
                TestConstants.HANDICAP_GROUP_ID,
                TestConstants.HANDICAP_GROUP_NAME,
                TestConstants.HANDICAP_GROUP_FACTOR
        );
    }

    public static HandicapGroupHibernateEntity fullHandicapGroup(
            String id,
            String name,
            BigDecimal factor
    ) {
        HandicapGroupHibernateEntity handicapGroup = new HandicapGroupHibernateEntity();
        handicapGroup.setId(id);
        handicapGroup.setName(name);
        handicapGroup.setFactor(factor);
        return handicapGroup;
    }

    public static HandicapGroupSetHibernateEntity fullHandicapGroupSet() {
        return fullHandicapGroupSet(
                TestConstants.HANDICAP_GROUP_SET_ID,
                TestConstants.HANDICAP_GROUP_SET_NAME,
                Sets.newHashSet(fullHandicapGroup())
        );
    }

    public static HandicapGroupSetHibernateEntity fullHandicapGroupSet(
            String id,
            String name,
            Set<HandicapGroupHibernateEntity> handicapGroups
    )  {
        HandicapGroupSetHibernateEntity handicapGroupSet = new HandicapGroupSetHibernateEntity();
        handicapGroupSet.setId(id);
        handicapGroupSet.setName(name);
        handicapGroupSet.setHandicapGroups(handicapGroups);
        return handicapGroupSet;
    }

    public static CompetitionGroupHibernateEntity fullCompetitionGroup() {
        return fullCompetitionGroup(
                TestConstants.COMPETITION_GROUP_ID,
                TestConstants.COMPETITION_GROUP_NAME,
                TestConstants.COMPETITION_GROUP_GROUPING,
                TestConstants.COMPETITION_GROUP_FACTOR,
                TestConstants.COMPETITION_GROUP_RESULT_TIME_TYPE.name()
        );

    }

    public static CompetitionGroupHibernateEntity fullCompetitionGroup(
            String id,
            String name,
            boolean grouping,
            BigDecimal handicapFactor,
            String resultTimeType
    ) {
        CompetitionGroupHibernateEntity competitionGroup = new CompetitionGroupHibernateEntity();
        competitionGroup.setId(id);
        competitionGroup.setName(name);
        competitionGroup.setGrouping(grouping);
        competitionGroup.setFactor(handicapFactor);
        competitionGroup.setResultTimeType(resultTimeType);
        return competitionGroup;
    }

    public static CompetitionGroupSetHibernateEntity fullCompetitionGroupSet() {
        return fullCompetitionGroupSet(
                TestConstants.COMPETITION_GROUP_SET_ID,
                TestConstants.COMPETITION_GROUP_SET_NAME,
                Sets.newHashSet(fullCompetitionGroup())
        );
    }

    public static CompetitionGroupSetHibernateEntity fullCompetitionGroupSet(
            String id,
            String name,
            Set<CompetitionGroupHibernateEntity> competitionGroups
    ) {
        CompetitionGroupSetHibernateEntity competitionGroupSet = new CompetitionGroupSetHibernateEntity();
        competitionGroupSet.setId(id);
        competitionGroupSet.setName(name);
        competitionGroupSet.setCompetitionGroups(competitionGroups);
        return competitionGroupSet;
    }

    public static RunHibernateEntity fullRun() {
        return RunHibernateEntity.builder()
                .id(TestConstants.RUN_ID)
                .event(fullEvent())
                .registration(fullRegistration())
                .sequence(TestConstants.RUN_SEQUENCE)
                .timestamp(TestConstants.RUN_TIMESTAMP)
                .rawTime(TestConstants.RUN_RAW_TIME)
                .cones(TestConstants.RUN_CONES)
                .didNotFinish(TestConstants.RUN_DID_NOT_FINISH)
                .disqualified(TestConstants.RUN_DISQUALIFIED)
                .rerun(TestConstants.RUN_RERUN)
                .competitive(TestConstants.RUN_COMPETITIVE)
                .build();
    }
}

File: service/src/test/java/org/coner/core/util/HibernateEntityTestUtils.kt
package org.coner.core.util

import com.google.common.collect.Sets
import org.coner.core.hibernate.entity.*
import java.math.BigDecimal
import java.time.Instant
import java.time.Year
import java.util.*

object HibernateEntityTestUtils {

    @JvmOverloads
    @JvmStatic
    fun fullPerson(
            id: String? = TestConstants.PERSON_ID,
            firstName: String = TestConstants.PERSON_FIRST_NAME,
            middleName: String = TestConstants.PERSON_MIDDLE_NAME,
            lastName: String = TestConstants.PERSON_LAST_NAME
    ): PersonHibernateEntity {
        val person = PersonHibernateEntity()
        person.id = id
        person.firstName = firstName
        person.middleName = middleName
        person.lastName = lastName
        return person
    }

    @JvmOverloads
    @JvmStatic
    fun fullCar(
            id: String? = TestConstants.CAR_ID,
            year: Year = TestConstants.CAR_YEAR,
            make: String = TestConstants.CAR_MAKE,
            model: String = TestConstants.CAR_MODEL,
            trim: String = TestConstants.CAR_TRIM,
            color: String = TestConstants.CAR_COLOR
    ): CarHibernateEntity {
        val car = CarHibernateEntity()
        car.id = id
        car.year = year
        car.make = make
        car.model = model
        car.trim = trim
        car.color = color
        return car
    }

    @JvmOverloads
    @JvmStatic
    fun fullEvent(
            id: String? = TestConstants.EVENT_ID,
            name: String = TestConstants.EVENT_NAME,
            date: Date = TestConstants.EVENT_DATE,
            handicapGroupSet: HandicapGroupSetHibernateEntity = fullHandicapGroupSet(),
            competitionGroupSet: CompetitionGroupSetHibernateEntity = fullCompetitionGroupSet(),
            maxRunsPerRegistration: Int = TestConstants.EVENT_MAX_RUNS_PER_REGISTRATION,
            current: Boolean = TestConstants.EVENT_CURRENT
    ): EventHibernateEntity {
        val event = EventHibernateEntity()
        event.id = id
        event.name = name
        event.date = date
        event.handicapGroupSet = handicapGroupSet
        event.competitionGroupSet = competitionGroupSet
        event.maxRunsPerRegistration = maxRunsPerRegistration
        event.isCurrent = current
        return event
    }

    @JvmOverloads
    @JvmStatic
    fun fullRegistration(
            id: String? = TestConstants.REGISTRATION_ID,
            person: PersonHibernateEntity = fullPerson(),
            car: CarHibernateEntity = fullCar(),
            event: EventHibernateEntity = fullEvent(),
            handicapGroup: HandicapGroupHibernateEntity = fullHandicapGroup(),
            competitionGroup: CompetitionGroupHibernateEntity = fullCompetitionGroup(),
            number: String = TestConstants.REGISTRATION_NUMBER,
            checkedIn: Boolean = TestConstants.REGISTRATION_CHECKED_IN
    ): RegistrationHibernateEntity {
        val registration = RegistrationHibernateEntity()
        registration.id = id
        registration.person = person
        registration.car = car
        registration.event = event
        registration.handicapGroup = handicapGroup
        registration.competitionGroup = competitionGroup
        registration.number = number
        registration.isCheckedIn = checkedIn
        return registration
    }

    @JvmOverloads
    @JvmStatic
    fun fullHandicapGroup(
            id: String? = TestConstants.HANDICAP_GROUP_ID,
            name: String = TestConstants.HANDICAP_GROUP_NAME,
            factor: BigDecimal = TestConstants.HANDICAP_GROUP_FACTOR
    ): HandicapGroupHibernateEntity {
        val handicapGroup = HandicapGroupHibernateEntity()
        handicapGroup.id = id
        handicapGroup.name = name
        handicapGroup.factor = factor
        return handicapGroup
    }

    @JvmOverloads
    @JvmStatic
    fun fullHandicapGroupSet(
            id: String? = TestConstants.HANDICAP_GROUP_SET_ID,
            name: String = TestConstants.HANDICAP_GROUP_SET_NAME,
            handicapGroups: Set<HandicapGroupHibernateEntity> = Sets.newHashSet(fullHandicapGroup())
    ): HandicapGroupSetHibernateEntity {
        val handicapGroupSet = HandicapGroupSetHibernateEntity()
        handicapGroupSet.id = id
        handicapGroupSet.name = name
        handicapGroupSet.handicapGroups = handicapGroups
        return handicapGroupSet
    }

    @JvmOverloads
    @JvmStatic
    fun fullCompetitionGroup(
            id: String? = TestConstants.COMPETITION_GROUP_ID,
            name: String = TestConstants.COMPETITION_GROUP_NAME,
            grouping: Boolean = TestConstants.COMPETITION_GROUP_GROUPING,
            handicapFactor: BigDecimal = TestConstants.COMPETITION_GROUP_FACTOR,
            resultTimeType: String = TestConstants.COMPETITION_GROUP_RESULT_TIME_TYPE.name
    ): CompetitionGroupHibernateEntity {
        val competitionGroup = CompetitionGroupHibernateEntity()
        competitionGroup.id = id
        competitionGroup.name = name
        competitionGroup.isGrouping = grouping
        competitionGroup.factor = handicapFactor
        competitionGroup.resultTimeType = resultTimeType
        return competitionGroup
    }

    @JvmOverloads
    @JvmStatic
    fun fullCompetitionGroupSet(
            id: String? = TestConstants.COMPETITION_GROUP_SET_ID,
            name: String = TestConstants.COMPETITION_GROUP_SET_NAME,
            competitionGroups: Set<CompetitionGroupHibernateEntity> = Sets.newHashSet(fullCompetitionGroup())
    ): CompetitionGroupSetHibernateEntity {
        val competitionGroupSet = CompetitionGroupSetHibernateEntity()
        competitionGroupSet.id = id
        competitionGroupSet.name = name
        competitionGroupSet.competitionGroups = competitionGroups
        return competitionGroupSet
    }

    @JvmOverloads
    @JvmStatic
    fun fullRun(
            id: String? = TestConstants.RUN_ID,
            event: EventHibernateEntity = fullEvent(),
            registration: RegistrationHibernateEntity = fullRegistration(),
            sequence: Int = TestConstants.RUN_SEQUENCE,
            timestamp: Instant = TestConstants.RUN_TIMESTAMP,
            rawTime: BigDecimal = TestConstants.RUN_RAW_TIME,
            cones: Int = TestConstants.RUN_CONES,
            didNotFinish: Boolean = TestConstants.RUN_DID_NOT_FINISH,
            disqualified: Boolean = TestConstants.RUN_DISQUALIFIED,
            rerun: Boolean = TestConstants.RUN_RERUN,
            competitive: Boolean = TestConstants.RUN_COMPETITIVE
    ) = RunHibernateEntity(
            id = id,
            event = event,
            registration = registration,
            sequence = sequence,
            timestamp = timestamp,
            rawTime = rawTime,
            cones = cones,
            didNotFinish = didNotFinish,
            disqualified = disqualified,
            rerun = rerun,
            competitive = competitive
    )
}
